### ISSUE TITLE
ENH: Implement matmul function

### ISSUE THREAD

**Issue by abalkin** [numpy/numpy#4464]
The @-operator PEP (gh-4351) describes two operations that would be a valuable addition to numpy independently of the main purpose of the PEP.

Most of the functionality is already available in various places (`np.dot`, `np.linalg.matrix_power`, etc.) but no single function conforms to the PEP.

**Comment by nicktimko** [comment]
Python 3.5.0a3 has added the `@` binary operator to the syntax, is there some NumPy branch available to test with? Naively trying to subclass from `np.array` fails because `np.ndarray` is actually the type and it gets really messy to recast everything (I don't think I can monkey-patch the class...)

**Comment by charris** [comment]
I had a toy implementation and it worked fine with 3.5 as far as I tested. I'm working on an 'official' version now for Numpy 1.10.

**Comment by abalkin** [comment]
@nicktimko - if you would like to contribute pure python code to this effort, try implementing `__matmul__` for masked arrays.  (See numpy.ma subpackage.)

**Comment by nicktimko** [comment]
What's the procedure for writing tests that involve new syntax? Just don't use it and manually call the method, i.e. `a.__matmul__(b)` instead of `a  @ b`?

**Comment by pv** [comment]
Check Python version, and use eval("a @ b")?

**Comment by njsmith** [comment]
Both! (Because we'd like the tests to actually exercise the code now, even
before we have 3.5 on our CI servers, but we'd also like to test that the
syntax actually works.)

**Comment by charris** [comment]
We would want to test both older an newer versions of Python. For the older
versions we would do something like the `operator.matmul`. Not sure what
the best way to test the new syntax is, maybe have a test install that is
version dependent?

**Comment by seberg** [comment]
Not sure if it is practical, but I guess you could also avoid writing "@" completly by doing something like this (not quite sure I got all the magic python does right):
```
try:
    from operator import matmul
except ImportError:
    def matmul(a, b):
         # subtypes have priority:
         reversed_first =  issubtype(type(b), type(a))
         if reversed_first:
              res = b.__rmatmul__(a)
         else:
              res = a.__matmul__(b)

         if res is not NotImplemented:
              return res
         # Try the opposite way around, since NotImplemented was returned:
         if reversed_first:
             return a.__matmul__(b)
         else:
             return b.__rmatmul__(a)
```

**Comment by nicktimko** [comment]
Can you capture `NotImplemented` as a return value from an unimplemented `__matmul__`? I thought only the comparisons (e.g. `__lt__`) give those, and for other magic operator methods it's just a proper AttributeError.

**Comment by seberg** [comment]
Oh, yeah....

**Comment by asmeurer** [comment]
Operators seem to return `NotImplemented` for me

```
In [1]: (1).__add__([])
Out[1]: NotImplemented
```

**Comment by argriffing** [comment]
> Operators seem to return `NotImplemented` for me

Maybe I'm not following correctly, but I thought the problem was something like this:
```python
>>> from operator import matmul
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name matmul
>>> (1).__add__([])
NotImplemented
>>> (1).__matmul__([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'int' object has no attribute '__matmul__'
```

**Comment by asmeurer** [comment]
Ah, I see. That is even in Python 3.5.

**Comment by nicktimko** [comment]
Would it work to directly dig out the method for some tests (`a_array.__matmul__(b_array)`, et al.), and do the syntax based on something like:

    matrix_op_syntax = True
    try:
        compile('x @ x', '<string>', 'eval')
    except SyntaxError:
        matrix_op_syntax = False

    ...

    def test_syntax_whatever(self):
        if not matrix_op_syntax:
            raise nose.SkipTest('Matrix operator syntax not supported')
        c = eval('a @ b')

**Comment by njsmith** [comment]
Sure, that would work.

**Comment by argriffing** [comment]
This is just a thought, but what about using a module that is not imported unless you know the Python version can deal with `a @ b`?  The idea would be that Pythons that don't understand the syntax wouldn't even look at the file, so you could avoid wrapping the tests in `eval`.

**Comment by charris** [comment]
On Mon, Apr 6, 2015 at 4:16 PM, Nick Timkovich <notifications@github.com>
wrote:

> Python 3.5.0a3 has added the @ binary operator to the syntax, is there
> some NumPy branch available to test with? Naively trying to subclass from
> np.array fails because np.ndarray is actually the type and it gets really
> messy to recast everything (I don't think I can monkey-patch the class...)
>
>
BTW, the class I used to experiment with is an attachment to this post
<http://article.gmane.org/gmane.comp.python.numeric.general/58500/match=matmul>
.

Chuck

**Comment by rgommers** [comment]
@charris maybe put it up as a gist for convenience?

**Comment by charris** [comment]
@nicktimko I've implemented `__matmul__` and `__rmatmul__` in #5878. The methods seem to work, but they do not get called by the `@` operator. Perhaps that is because ndarray is a type defined in C.

**Comment by shoyer** [comment]
This was fixed by #5878
